src/coreclr/jit/block.h

```diff
@@ -662,13 +662,13 @@ struct BasicBlock : private LIR::Range
 
     bool HasJump() const
     {
+        assert(KindIs(BBJ_ALWAYS, BBJ_CALLFINALLY, BBJ_COND, BBJ_EHCATCHRET, BBJ_EHFILTERRET, BBJ_LEAVE));
         return (bbJumpDest != nullptr);
     }
```

Comment:
What is the point of `HasJump()`? It seems like a check for whether the block is in some half-constructed state, since I would expect the block kinds above always to have valid `bbJumpDest`. It does not really make sense for me to have this function with the assert that is added here.
An assert that would make more sense to me would be:
```
assert(KindIs(BBJ_ALWAYS, BBJ_CALLFINALLY, BBJ_COND, BBJ_EHCATCHRET, BBJ_EHFILTERRET, BBJ_LEAVE) == (bbJumpDest != nullptr))
```

Reply:
Ah, I see that @BruceForstall's feedback in #94239 was about this being a union field.

I would like to understand when it is expected that these jump kinds have a null `bbJumpDest`. Maybe that happens only for a very short while when we're initializing the flow graph? In that case, consider renaming this function to `HasInitializedJumpDest()` or something of the sort, to make that clear. I think it is a bit confusing currently, as it makes it seem like an expected situation that some of these blocks do not have a valid `bbJumpDest`, which should not be possible.

Reply:
> I would like to understand when it is expected that these jump kinds have a null bbJumpDest. Maybe that happens only for a very short while when we're initializing the flow graph?

Yes, there are a few cases where we have to create a block that requires a jump target without knowing the target yet. For example, in `Compiler::fgMakeBasicBlocks`, we create blocks only knowing the offset of their jump targets, and set the actual target block later in `Compiler::fgLinkBasicBlocks`. The tougher cases are during importing, where we sometimes create a `BBJ_ALWAYS` or `BBJ_CALLFINALLY` without setting its jump target in a loop, and then create and set the jump target in a later iteration of the loop. There are a few other spots where we do this pattern, but I can't remember off the top of my head...

Your first suggested assert seems to best communicate what `HasJump` should do; for block kinds that should never read `bbJumpDest`, `bbJumpDest` should not be set, else `bbJumpDest` must be set (i.e. we shouldn't be in some half-constructed state) before performing any conditional logic based on whether the block has a jump or not.